### PR TITLE
Intercept exceptions and provide a human readable message for the users

### PR DESCRIPTION
**Pull request by zalun**
Exceptions are intercepted in the `cinnabar.util.run` method.
If an exception is raised with a message, it is printed out.

A "traceback" check is introduced to provide an option to see the
traceback.

A new module `cinnabar/exceptions.py` is added to define all specific
exceptions.

`Abort` is defined as a base exception for normal command aborts.

`HelperClosed` is now a RuntimeError.

### CODE REVIEW

File: cinnabar/util.py [hunk 1]
```diff
@@ -608,6 +609,20 @@ def run(func):
         reporter = MemoryReporter()
     try:
         retcode = func(sys.argv[1:])
+    except Exception as e:
+        # Catch all exceptions and provide a nice message
+        retcode = 1
+        sys.stderr.write(
+            'Sorry. The cinnabar extension exited because of an error.\n')
```
Comment (glandium) [inline cinnabar/util.py]: "The cinnabar extension" works when you do something like `git pull`, but when the command you run is `git cinnabar $command`, that's weird. I think we can skip this message entirely.

File: cinnabar/util.py [hunk 2]
```diff
@@ -608,6 +609,20 @@ def run(func):
...
+        if check_enabled('traceback'):
+            traceback.print_exc()
+        else:
+            if hasattr(e, 'message') and e.message:
+                sys.stderr.write('{}\n'.format(e.message))
+
+            sys.stderr.write(
+                'Please add the `cinnabar.check = traceback` setting to the '
```
Comment (glandium) [inline cinnabar/util.py]: Mmmm. I'm not sure advertising changing the git config is great. Moreover, if you edit your git configuration manually, you don't really write it this way. It's probably better to indicate how to do it for one command, which is to use `git -c cinnabar.check=traceback <command>`.

File: .gitignore [hunk 3]
```diff
@@ -1,2 +1,3 @@
 git-cinnabar-helper
 *.patched.c
+*.pyc
```
Comment (glandium) [inline .gitignore]: Can you do a separate PR for this?

File: cinnabar/util.py (hunk 1 above)
Comment (zalun) [reply cinnabar/util.py]: So - maybe "ERROR: {}".format(e.message) ?

File: cinnabar/util.py [hunk 4]
```diff
@@ -608,6 +609,19 @@ def run(func):
...
     try:
         retcode = func(sys.argv[1:])
+    except Exception as e:
+        # Catch all exceptions and provide a nice message
+        retcode = 1
+        if check_enabled('traceback'):
+            traceback.print_exc()
+        else:
+            if hasattr(e, 'message') and e.message:
+                logger.error(e.message)
```
[the remark this answers is not in the document]
Comment (zalun) [reply cinnabar/util.py]: The `NoHelperException` and `HelperClosedException` have no message. They do log the error though. Do you think they should `print_exc()` even if no traceback requested?

**Comment by glandium** [comment]
I did small adjustments to the run function. Waiting for the CI results before merging.
